File: java/src/db/PersonajeDAO.java
package db;

import model.Personaje;
import model.Guerrero;
import model.Mago;
import model.Arquero;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

import java.util.ArrayList;

public class PersonajeDAO {

    private Conexion conexion;

    // constructor
    public PersonajeDAO() {

        conexion = new Conexion();
    }

    // INSERT
    public void insertarPersonaje(
            Personaje personaje
    ) {

        try{

            Connection con =
                    conexion.conectar();

            String sql =
                    "INSERT INTO personaje(nombre,nivel)"
                    + " VALUES(?,?)";

            PreparedStatement ps =
                    con.prepareStatement(sql);

            ps.setString(
                    1,
                    personaje.getNombre()
            );

            ps.setInt(
                    2,
                    personaje.getNivel()
            );

            ps.executeUpdate();

        }catch(Exception e){

            System.out.println(
                    e.getMessage()
            );
        }
    }

    // SELECT
    public ArrayList<Personaje>
    listarPersonajes(){

        ArrayList<Personaje> lista =
                new ArrayList<>();

        try{

            Connection con =
                    conexion.conectar();

            String sql =
                    "SELECT * FROM personaje";

            PreparedStatement ps =
                    con.prepareStatement(sql);

            ResultSet rs =
                    ps.executeQuery();

            while(rs.next()){

                String clase =
                        rs.getString("clase");

                Personaje p;

                if(clase.equals("Guerrero")){
                    p = new Guerrero(
                            rs.getInt("id"),
                            rs.getString("nombre"),
                            rs.getInt("nivel")
                    );
                }else if(clase.equals("Mago")){
                    p = new Mago(
                            rs.getInt("id"),
                            rs.getString("nombre"),
                            rs.getInt("nivel")
                    );
                }else{
                    p = new Arquero(
                            rs.getInt("id"),
                            rs.getString("nombre"),
                            rs.getInt("nivel")
                    );
                }

                lista.add(p);
            }

        }catch(Exception e){

            System.out.println(
                    e.getMessage()
            );
        }

        return lista;
    }

    // Buscar por nombre
    public Personaje buscarPorNombre(
            String nombre
    ){

        Personaje personaje = null;

        try{

            Connection con =
                    conexion.conectar();

            String sql =
                    "SELECT * FROM personaje "
                    + "WHERE nombre=?";

            PreparedStatement ps =
                    con.prepareStatement(sql);

            ps.setString(1, nombre);

            ResultSet rs =
                    ps.executeQuery();

            if(rs.next()){

                String clase =
                        rs.getString("clase");

                if(clase.equals("Guerrero")){
                    personaje = new Guerrero(
                            rs.getInt("id"),
                            rs.getString("nombre"),
                            rs.getInt("nivel")
                    );
                }else if(clase.equals("Mago")){
                    personaje = new Mago(
                            rs.getInt("id"),
                            rs.getString("nombre"),
                            rs.getInt("nivel")
                    );
                }else{
                    personaje = new Arquero(
                            rs.getInt("id"),
                            rs.getString("nombre"),
                            rs.getInt("nivel")
                    );
                }
            }

        }catch(Exception e){

            System.out.println(
                    e.getMessage()
            );
        }

        return personaje;
    }

    // UPDATE
    public void actualizarNivel(
            int id,
            int nivel
    ){

        try{

            Connection con =
                    conexion.conectar();

            String sql =
                    "UPDATE personaje "
                    + "SET nivel=? "
                    + "WHERE id=?";

            PreparedStatement ps =
                    con.prepareStatement(sql);

            ps.setInt(1, nivel);
            ps.setInt(2, id);

            ps.executeUpdate();

        }catch(Exception e){

            System.out.println(
                    e.getMessage()
            );
        }
    }

    // DELETE
    public void eliminarPersonaje(
            int id
    ){

        try{

            Connection con =
                    conexion.conectar();

            String sql =
                    "DELETE FROM personaje "
                    + "WHERE id=?";

            PreparedStatement ps =
                    con.prepareStatement(sql);

            ps.setInt(1, id);

            ps.executeUpdate();

        }catch(Exception e){

            System.out.println(
                    e.getMessage()
            );
        }
    }
}

File: python/db/personaje_dao.py
from db.conexion import Conexion

from model.guerrero import Guerrero
from model.mago import Mago
from model.arquero import Arquero

class PersonajeDAO:

    # constructor
    def __init__(self):

        self.conexion = Conexion()

    # INSERT
    def insertar_personaje(
            self,
            personaje
    ):

        try:

            con = self.conexion.conectar()

            cursor = con.cursor()

            sql = """
            INSERT INTO personaje(
            nombre,
            nivel,
            clase
            )
            VALUES(%s,%s,%s)
            """

            valores = (
                personaje.nombre,
                personaje.nivel,
                personaje.__class__.__name__
            )

            cursor.execute(sql, valores)

            con.commit()

        except Exception as e:

            print(e)

    # SELECT
    def listar_personajes(self):

        lista = []

        try:

            con = self.conexion.conectar()

            cursor = con.cursor()

            sql = "SELECT * FROM personaje"

            cursor.execute(sql)

            resultados = cursor.fetchall()

            for fila in resultados:

                clase = fila[3]

                if clase == "Guerrero":

                    personaje = Guerrero(
                        fila[0],
                        fila[1],
                        fila[2]
                    )

                elif clase == "Mago":

                    personaje = Mago(
                        fila[0],
                        fila[1],
                        fila[2]
                    )

                else:

                    personaje = Arquero(
                        fila[0],
                        fila[1],
                        fila[2]
                    )

                lista.append(personaje)

        except Exception as e:

            print(e)

        return lista

    # Buscar por nombre
    def buscar_por_nombre(
            self,
            nombre
    ):

        personaje = None

        try:

            con = self.conexion.conectar()

            cursor = con.cursor()

            sql = """
            SELECT * FROM personaje
            WHERE nombre=%s
            """

            cursor.execute(sql, (nombre,))

            fila = cursor.fetchone()

            if fila:

                clase = fila[3]

                if clase == "Guerrero":

                    personaje = Guerrero(
                        fila[0],
                        fila[1],
                        fila[2]
                    )

                elif clase == "Mago":

                    personaje = Mago(
                        fila[0],
                        fila[1],
                        fila[2]
                    )

                else:

                    personaje = Arquero(
                        fila[0],
                        fila[1],
                        fila[2]
                    )

        except Exception as e:

            print(e)

        return personaje

    # UPDATE
    def actualizar_nivel(
            self,
            id,
            nivel
    ):

        try:

            con = self.conexion.conectar()

            cursor = con.cursor()

            sql = """
            UPDATE personaje
            SET nivel=%s
            WHERE id=%s
            """

            cursor.execute(
                sql,
                (nivel, id)
            )

            con.commit()

        except Exception as e:

            print(e)

    # DELETE
    def eliminar_personaje(
            self,
            id
    ):

        try:

            con = self.conexion.conectar()

            cursor = con.cursor()

            sql = """
            DELETE FROM personaje
            WHERE id=%s
            """

            cursor.execute(
                sql,
                (id,)
            )

            con.commit()

        except Exception as e:

            print(e)